File: py/Day09.py
from collections import defaultdict
import time



# Helper functions
def get_last_index_of_non_dot(blocks, last_found):
    for i in range(last_found, -1, -1):
        if blocks[i] != -1:
            return i
    return -1

def get_last_index_of_id(blocks, block_id):
    for i in range(len(blocks) - 1, -1, -1):
        if blocks[i] == block_id:
            return i
    return -1

def get_first_index_of_dot(blocks, first_found):
    for i in range(first_found, len(blocks)):
        if blocks[i] == -1:
            return i
    return -1

def get_first_index_of_n_dot(blocks, length, end):
    for i in range(0, end - length + 1):
        if blocks[i] == -1 and i + length <= len(blocks):
            if all(blocks[i + j] == -1 for j in range(length)):
                return i
    return -1

def process_file(file_content):
    # Parse the input
    disk_map = file_content.strip()
    
    # Validate input length
    if len(disk_map) % 2 != 0:
        disk_map = disk_map+"0"

    # Decode the disk map into individual blocks
    ids = []
    ids_map = {}
    blocks_list1 = []
    blocks_list2 = []

    for i in range(0, len(disk_map), 2):
        file_length = int(disk_map[i])
        free_length = int(disk_map[i + 1])
        block_id = i // 2

        blocks_list1.extend([block_id] * file_length)
        blocks_list1.extend([-1] * free_length)

        if block_id not in ids:
            ids.append(block_id)
        if block_id not in ids_map:
            ids_map[block_id] = file_length

    blocks_list2 = blocks_list1[:]

    # Simulate the compaction process
    write_index = get_first_index_of_dot(blocks_list1, 0)
    read_index = get_last_index_of_non_dot(blocks_list1, len(blocks_list1) - 1)

    while write_index < read_index:
        blocks_list1[write_index], blocks_list1[read_index] = blocks_list1[read_index], -1
        write_index = get_first_index_of_dot(blocks_list1, write_index)
        read_index = get_last_index_of_non_dot(blocks_list1, read_index)

    # Calculate the checksum for part 1
    result_part1 = sum(i * val for i, val in enumerate(blocks_list1) if val != -1)

    # Reverse the order of IDs for part 2
    ids.reverse()
    for current_id in ids:
        n = ids_map[current_id]
        end = get_last_index_of_id(blocks_list2, current_id)
        write_index = get_first_index_of_n_dot(blocks_list2, n, end)

        if write_index >= 0:
            for i in range (n):
                blocks_list2.remove(current_id)
            for i in range(write_index, write_index+n):
                o = blocks_list2.pop(write_index)
                blocks_list2.insert(end-n, o) 
            
            for i in range (n):
                blocks_list2.insert(write_index,current_id)
            
    # Calculate the checksum for part 2
    result_part2 = sum(i * val for i, val in enumerate(blocks_list2) if val != -1)

    return result_part1, result_part2

def main():
    time_start = int(time.time() * 1000)
    file_path = r'..\input_2024\day09_data.txt'

    try:
        with open(file_path, "r") as file:
            file_content = file.read()

    except FileNotFoundError:
        print("Error: File not found.")
        return
    
    result_part_1, result_part_2 = process_file(file_content)

    print("result part 1:", result_part_1)
    print("result part 2:", result_part_2)

    time_end = int(time.time() * 1000)
    print(f"Execution time: {time_end - time_start} msec")

if __name__ == "__main__":
    main()

File: src/AoC2024/Day09.java
package AoC2024;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import AoC2022.Day;

public class Day09 extends Day {

	public Day09() {
		super();
		LOG_LEVEL = 1;
		INPUT_REAL = true;
	}
	
	private int getLastIndexOfNonDot(List<Integer> blocks, int lastFound) {
		for (int i = lastFound; i > 0 ; i--) {
			if(blocks.get(i).intValue()!=-1) return i;
		}
		return -1;
	}
	
	private int getLastIndexOfID(List<Integer> blocks, Integer id) {
		for (int i = blocks.size()-1; i > 0 ; i--) {
			if(blocks.get(i).intValue()==id.intValue()) return i;
		}
		return -1;
	}
	
	private int getFirstIndexOfDot(List<Integer> blocks, int firstFound) {
		for (int i = firstFound; i < blocks.size() ; i++) {
			if(blocks.get(i).intValue()==-1) return i;
		}
		return -1;
	}
	
	private int getFirstIndexOfNDot(List<Integer> blocks,int len, int end) {
		for (int i = 0; i < end-len ; i++) {
			if(blocks.get(i).intValue()==-1 && i+len < blocks.size()) {
				boolean enoughDots = true;
				for (int j = 1; j < len; j++) {
					if(blocks.get(i+j).intValue()!=-1) {
						enoughDots = false;
						break;
					}
				}
				if(enoughDots) return i;
			}
		}
		return -1;
	}

	public void processFile() {
	    // Parse the input from the sb variable
	    String diskMap = sb.toString().trim();

	    // Validate the input length
	    if (diskMap.length() % 2 != 0) {
	        //System.err.println("Invalid input: Length of disk map is not even. Adding 0 to the end.");
	        diskMap=diskMap+"0";
	    }
	    
	    List<Integer> ids = new ArrayList<Integer>();
	    HashMap<Integer,Integer> ids_map = new HashMap<Integer,Integer>();

	    // Step 1: Decode the disk map into individual blocks
	    List<Integer> blocks_list1 = new ArrayList<Integer>();
	    List<Integer> blocks_list2 = new ArrayList<Integer>();
	    for (int i = 0; i < diskMap.length(); i += 2) {
	        int fileLength = Character.getNumericValue(diskMap.charAt(i));
	        int freeLength = Character.getNumericValue(diskMap.charAt(i + 1));
	        
	        // Append file blocks
	        for (int j = 0; j < fileLength; j++) {
	        	Integer id = (i / 2);
	        	blocks_list1.add(id);
	        	if(!ids.contains(id)) ids.add(id);
	        	if(!ids_map.containsKey(id)) ids_map.put(id,fileLength);
	        	
	        }
	        // Append free space
	        for (int j = 0; j < freeLength; j++) {
	        	blocks_list1.add(-1);
	        }
	    }
	    blocks_list2.addAll(blocks_list1);
	    
	    // Step 2: Simulate the compaction process
	    int writeIndex = getFirstIndexOfDot(blocks_list1,0);
	    int readIndex = getLastIndexOfNonDot(blocks_list1,blocks_list1.size()-1);
	    
	    while(writeIndex<readIndex) {
	    	blocks_list1.remove(writeIndex);
	    	blocks_list1.add(writeIndex, blocks_list1.get(readIndex-1));
	    	blocks_list1.remove(readIndex);
	    	blocks_list1.add(readIndex, -1);
	    	
	    	writeIndex = getFirstIndexOfDot(blocks_list1,writeIndex);
	    	readIndex = getLastIndexOfNonDot(blocks_list1,readIndex);
	    }

	    // Step 3: Calculate the checksum
	    long result_part1 = 0;
	    for (int i = 0; i < blocks_list1.size(); i++) {
	        if (blocks_list1.get(i).intValue()!=-1) {
	        	result_part1 += i * blocks_list1.get(i).intValue();
	        }
	    }

	    // Output the result for part 1
	    System.out.println("result part1: " + result_part1); // Correct: 6258319840548 1928
	    
	    Collections.reverse(ids);
	    for (Integer currentID : ids) {
	    	int n = ids_map.get(currentID).intValue();
	    	
	    	int end = getLastIndexOfID(blocks_list2,currentID);
		    
		    writeIndex = getFirstIndexOfNDot(blocks_list2,n,end);
		    
		    if(writeIndex >= 0) {
		    	
		    	blocks_list2.removeAll(Collections.singletonList(currentID));
		    	for (int i = writeIndex; i < writeIndex+n; i++) {
		    		Integer o = blocks_list2.remove(writeIndex);
		    		blocks_list2.add(end-n,o);
				}
//		    	for (int i = 0; i < n; i++) {
//		    		blocks_list2.add(writeIndex,currentID);
//				}
		    	
		    	blocks_list2.addAll(writeIndex,Stream.generate(() -> currentID).limit(n).collect(Collectors.toCollection(ArrayList<Integer>::new)));
		    } 
	    }
	    
	    // Step 3: Calculate the checksum
	    long result_part2 = 0;
	    for (int i = 0; i < blocks_list2.size(); i++) {
	        if (blocks_list2.get(i).intValue()!=-1) {
	        	result_part2 += i * blocks_list2.get(i).intValue();
	        }
	    }
	    
	    // Placeholder for part 2
	    System.out.println("result part2: " + result_part2); // Correct: 6286182965311 2858
	}



	public static void main(String[] args) {
		long time_start = System.currentTimeMillis();
		Day09 day = new Day09();
		day.readFile();
		day.processFile();
		long time_end = System.currentTimeMillis();
		day.logln("Execution time: "+(time_end-time_start)+" msec");
		day.flush();
	}

}
